### PR TITLE
CR

### PR DESCRIPTION
**Pull request by maxponte**

### CODE REVIEW

File: has_key.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package selector
+
+// HasKey returns if a label set has a given key.
+type HasKey string
+
+// Matches returns the selector result.
+func (hk HasKey) Matches(labels Labels) bool {
+	if _, hasKey := labels[string(hk)]; hasKey {
+		return true
```
Comment (maxponte) [inline has_key.go]: just return hasKey

File: lexer.go [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+	OpDoubleEquals = "=="
+	// OpNotEquals is an operator.
+	OpNotEquals = "!="
+	// OpIn is an operator.
+	OpIn = "in"
+	// OpNotIn is an operator.
+	OpNotIn = "notin"
+)
+
+// Lexer is the working engine of the semantic extraction for a selector.
```
Comment (maxponte) [inline lexer.go]: confused by this definition

File: lexer.go [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+type Lexer struct {
+	// s stores the string to be tokenized
+	s string
+	// pos is the position currently tokenized
+	pos int
+	// m is an optional mark
+	m int
+}
+
+// Lex does the actual parsing.
```
Comment (maxponte) [inline lexer.go]: isn't it like, you lex the input, it produces a token stream, you parse that, as separate stages

File: lexer.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+	}
+}
+
+// isWhitespace returns true if the rune is a space, tab, or newline.
+func (l *Lexer) isWhitespace(ch rune) bool {
+	return ch == Space || ch == Tab || ch == CarriageReturn || ch == NewLine
+}
+
+// isSpecialSymbol returns if the ch can be a token.
+func (l *Lexer) isSpecialSymbol(ch rune) bool {
```
Comment (maxponte) [inline lexer.go]: inconsistent naming

File: lexer.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+
+// advance moves the cursor forward one character.
+func (l *Lexer) advance() {
+	if l.pos < len(l.s) {
+		_, width := utf8.DecodeRuneInString(l.s[l.pos:])
+		l.pos += width
+	}
+}
+
+// unread moves the cursor back a character.
```
Comment (maxponte) [inline lexer.go]: unread => prev

File: lexer.go [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+func (l *Lexer) read() (r rune) {
+	var width int
+	if l.pos < len(l.s) {
+		r, width = utf8.DecodeRuneInString(l.s[l.pos:])
+		l.pos += width
+	}
+	return r
+}
+
+// current returns the byte a the current position.
```
Comment (maxponte) [inline lexer.go]: cmt typo

File: parse.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+//  (1) Inclusion - " in " - denotes that the KEY exists and is equal to any of the
+//      VALUEs in its requirement
+//  (2) Exclusion - " notin " - denotes that the KEY is not equal to any
+//      of the VALUEs in its requirement or does not exist
+//  (3) The empty string is a valid VALUE
+//  (4) A requirement with just a KEY - as in "y" above - denotes that
+//      the KEY exists and can be any VALUE.
+//  (5) A requirement with just !KEY requires that the KEY not exist.
+//
+func Parse(query string) (Selector, error) {
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply parse.go]: yeah that was an iteration casualty.

File: has_key.go (hunk 1 above)
Comment (wcharczuk) [reply has_key.go]: 👍

File: lexer.go (hunk 2 above)
Comment (wcharczuk) [reply lexer.go]: 👍

File: lexer.go (hunk 3 above)
Comment (wcharczuk) [reply lexer.go]: that is usually how a lexer works, yes, what we have now is really a parser.

**Pull request closed without merging**